archivebox/index/sql.py

```diff
@@ -21,29 +22,35 @@ def parse_sql_main_index(out_dir: str=OUTPUT_DIR) -> Iterator[Link]:
     )
 
 @enforce_types
-def remove_from_sql_main_index(links: List[Link], out_dir: str=OUTPUT_DIR) -> None:
+def remove_from_sql_main_index(snapshots: QuerySet, out_dir: str=OUTPUT_DIR) -> None:
     setup_django(out_dir, check_db=True)
-    from core.models import Snapshot
     from django.db import transaction
 
     with transaction.atomic():
-        for link in links:
-            Snapshot.objects.filter(url=link.url).delete()
+        snapshots.delete()
+
+@enforce_types
+def write_link_to_sql_index(link: Link):
+    from core.models import Snapshot
+    info = {k: v for k, v in link._asdict().items() if k in Snapshot.keys}
+    try:
+        info["timestamp"] = Snapshot.objects.get(url=link.url).timestamp
+    except Snapshot.DoesNotExist:
```

Comment:
why does it increment int he case where it *doesn't* exist? shouldn't it increment in the case where it *does* already exist?

Reply:
There was a case (the one that was causing issues for users) where the url didn't exist, but the assigned timestamp was one that was already assigned to something else, causing a collision when it was assigned. In this case, if the url is not created yet, the following block tries to find (in a naive way) a suitable timestamp that does not cause a collision in the index so the object can be created.

Reply:
ah, got it